File: coolrmi/hu.qgears.coolrmi/src/hu/qgears/coolrmi/multiplexer/SocketMultiplexerDatagram.java
package hu.qgears.coolrmi.multiplexer;

import java.io.EOFException;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.Serializable;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

/**
 * Piece of a message on a multiplexed
 * message stream.
 * @author rizsi
 *
 */
public class SocketMultiplexerDatagram implements Serializable {
	private static final long serialVersionUID = 1L;
	private byte[] content;

	public SocketMultiplexerDatagram(byte[] data) {
		this.content=data;
	}

	public byte[] getContent() {
		return content;
	}
	public void setContent(byte[] content) {
		this.content = content;
	}
	@Override
	public String toString() {
		return "datagram: "+content.length;
	}
	public static SocketMultiplexerDatagram readFromStream(InputStream is) throws IOException {
		byte[] header=new byte[4];
		readAll(is, header);
		ByteBuffer bb=ByteBuffer.wrap(header);
		bb.order(ByteOrder.LITTLE_ENDIAN);
		int length=bb.getInt();

		byte[] data=new byte[length];
		readAll(is,  data);
		return new SocketMultiplexerDatagram(data);
	}
	private static void readAll(InputStream is, byte[] header) throws IOException {
		int at=0;
		while(at<header.length)
		{
			int n=is.read(header, at, header.length-at);
			if(n<0)
			{
				throw new EOFException();
			}
			at+=n;
		}
	}
	public void writeToStream(OutputStream os) throws IOException {
		byte[] header=new byte[4];
		ByteBuffer bb=ByteBuffer.wrap(header);
		bb.order(ByteOrder.LITTLE_ENDIAN);
		bb.putInt(content.length);
		os.write(header);
		os.write(content);
	}
}

File: coolrmi/CoolRMI.Net/Multiplexer/SocketMultiplexerDatagram.cs
using System;
using System.IO;

namespace CoolRMI.Net.Multiplexer
{
    public class SocketMultiplexerDatagram
    {
        public byte[] Content { get; set; }

        public SocketMultiplexerDatagram(byte[] content)
        {
            Content = content;
        }

        public override string ToString()
        {
            return "datagram: " + Content.Length;
        }

        public static SocketMultiplexerDatagram ReadFromStream(Stream stream)
        {
            var buf = new byte[4];
            ReadAll(stream, buf);
            if (!BitConverter.IsLittleEndian) Array.Reverse(buf);
            var len = BitConverter.ToInt32(buf, 0);

            buf = new byte[len];
            ReadAll(stream, buf);
            return new SocketMultiplexerDatagram(buf);
        }

        public void WriteToStream(Stream stream)
        {
            var buf = BitConverter.GetBytes(Content.Length);
            if (!BitConverter.IsLittleEndian) Array.Reverse(buf);
            stream.Write(buf, 0, 4);
            stream.Write(Content, 0, Content.Length);
        }

        private static void ReadAll(Stream stream, byte[] buf)
        {
            var at = 0;
            while (at < buf.Length)
            {
                var n = stream.Read(buf, at, buf.Length - at);
                if (n == 0)
                {
                    throw new EndOfStreamException();
                }
                at += n;
            }
        }
    }
}
